src/clearwater_modules/tsm/processes.py

```diff
@@ -327,40 +329,26 @@ def ri_number(
     )
 
 
-@numba.njit
-def ri_function(ri_number: float) -> float:
+def ri_function(ri_number: xr.DataArray) -> xr.DataArray:
     """Calculates the Richardson Function from the Richardson Number.
     Richardson Number:
         Unstable: 0.01 >= ri_function
         Stable: 0.01 <= ri_function < 2
         Neutral: -0.01 <  ri_function < 0.01
     """
-    # TODO: figure out how to make this work
-    return ri_number ** 2
-    # Set bounds
-    if ri_number > 2.0:
-        ri_number = 2.0
-    elif ri_number < -1.0:
-        ri_number = -1.0
-
-    if ri_number < 0.0:
-        if ri_number >= - 0.01:
-            # neutral
-            return 1.0
-        else:
-            # unstable
-            return (1.0 - 22.0 * ri_number)**0.80
-    else:
-        if ri_number <= 0.01:
-            # neutral
-            return 1.0
-        else:
-            # stable
-            return (1.0 + 34.0 * ri_number)**(-0.80)
+    warnings.filterwarnings("ignore", category=RuntimeWarning)
+    da: xr.DataArray = xr.where(ri_number > 2.0, 2.0, 
+        xr.where(ri_number < -1.0, -1.0,
+        xr.where((ri_number < 0.0) & (ri_number >= -0.01), 1.0,
+        xr.where(ri_number < -0.01, (1.0 - 22.0 * ri_number)**0.80, 
+        xr.where((ri_number >= 0.0) & (ri_number <= 0.01), 1.0, (1.0 + 34.0 * ri_number)**(-0.80)
+    )))))
```

Comment:
These nested `xr.where()` statements seem inefficient and hard to understand. 

Also, from what I can tell, the new code doesn't capture the same results from the original code, which itself has added complication over the simple statements made in the docstring. Can we go back to the source of this information and rewrite this from scratch?

Is there an xarray version of [`numpy.select()`](https://numpy.org/doc/stable/reference/generated/numpy.select.html), which would clearly encapsulate the 3 conditions described in the docstring? If so, that would be ideal.

Reply:
From @PeterKlaver:

> Originally, there were two separate steps, but it appears that they have been combined into one nested xr.where.
>  
> The first step sets bounds on the value of R – anything greater than 2 is set to 2, anything less than -1 is set to -1. The second step applies the value 1 to numbers between -0.01 and 0.01, and uses the functions outside of those ranges, including the ones that were capped at either 2 or -1. If I’m understanding the new code correctly, if it encounters ri_number greater than 2, it just sets it to 2 and moves on instead of going on to return a value based on (1.0 + 34.0 * ri_number)**(-0.80) . Same for values less than -1, which should be calculated from (1.0 - 22.0 * ri_number)**0.80.
>  
> Maybe this would accomplish the same thing:
>  
> ```python
> da: xr.DataArray = xr.where(ri_number > 2.0, (1.0 + 34.0*2.0)**(-0.80),
>         xr.where(ri_number < -1.0,  (1.0 – 22.0*-1.0)**(-0.80),
>         xr.where((ri_number < 0.0) & (ri_number >= -0.01), 1.0,
>         xr.where(ri_number < -0.01, (1.0 - 22.0 * ri_number)**0.80,
>         xr.where((ri_number >= 0.0) & (ri_number <= 0.01), 1.0, (1.0 + 34.0 * ri_number)**(-0.80)
>     )))))
> ```
>  
> Do you see what I’m seeing? Or am I missing something?

I short, you noticed the same thing as me, that two steps got conflated into one, with different results.  I also felt that the nesting of the where statements could be streamlined to make it easier to understand.

Reply:
@PeterKlaver, Thanks for your review and for drafting a rewrite of the function. You noticed the same thing as me, that two steps got conflated into one, with different results as you described. Thanks.

Your rewrite looks to better follow the original code and I would guess gives the same result (but we should confirm). 

I also felt that the original code was overly complicated and inefficient, and I think we could get the same results with fewer `xr.where()` statements that would also make it easier to understand.  For that reason, I was thinking to just look at the docstring for guidance, but it seems to me that there's a typo there. Shouldn't "Unstable: 2 <= ri_function", otherwise there is an interval (0.01 - 2.0) that is labeled as both Stable and Unstable, depending on the order of operations. That seems like an error in the original code and Docstring.

@PeterKlaver, is it obvious to you where all this came from?
We could also ask @BillyEJohnson or @drzhangus, or look in their reports describing TSM:

- Zhang, Zhonglong and Billy E. Johnson. 2016. Aquatic nutrient simulation modules (NSMs) developed for hydrologic and hydraulic models. Vicksburg, MS: Environmental Laboratory, U. S. Army Engineer Research and Development Center (ERDC). Ecosystem Management and Restoration Research Program (EMRRP). ERDC/EL Technical Report 16-1. https://hdl.handle.net/11681/10112

Reply:
I've looked at the above-referenced report. It's a bit awkward, but essentially a negative Richardson number is unstable and a positive R is stable, with R = 0 being neutral. What they've done is define a narrow band around zero (-0.01 to 0.01) to be "neutral". The "unstable" band if for R from -1.0 to -0.01, and the "stable" band is for R from 0.01 to 2.0. Outside of those ranges the value is constant, approximately equal to the appropriate function evaluated at the limit of the range. It ends up looking like a sigmoidal function, dropping from 12.3 to 0.03. This makes sense to me in the current application, which expresses the effect of wind speed on sensible heat flux: large for "unstable" conditions, dropping to a small number for stable conditions.

I'll look a level deeper to try to confirm the source of the equations, but I think the intent is clear. Seems to me that some kind of "where" construct is appropriate here, we have to apply different approaches depending on the value of ri_number. Of course the trick is specify something for all possible values, but not have any redundancy.

Reply:
If we need to get Zhong involved then use the following e-mail address: zz3@pdx.edu. Zhong was the original developer of this module with Todd Steissberg converting it into Python.

Reply:
There is not an xarray version of `numpy.select`, but I could just use the `numpy` function to update the data in place. Let me know when you land on a decision for what logic should be implemented.

Reply:
Switched to numpy select via commit 4ed6c5e6e8cf0224e40fd00a31db437b5920d3bc